core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -58,6 +57,12 @@ class Partition(val topic: String,
   @volatile var leaderReplicaIdOpt: Option[Int] = None
   @volatile var inSyncReplicas: Set[Replica] = Set.empty[Replica]
 
+  // The low watermark offset value, kept only if the local replica is the partition leader
+  // It is only used by leader broker to decide when DeleteRecordsRequest is satisfied. Its value is minimum logStartOffset of all live replicas
+  // It can be updated when the leader broker receives either FetchRequest or DeleteRecordsRequest.
+  // lowWatermark should always be accessed through updateAndGetLowWatermarkIfLocal() because we may not update lowWatermark when log segment is deleted.
+  @volatile private[this] var lowWatermarkIfLocal = INVALID_LOW_WATERMARK
```

Comment:
That is a good idea. I have updated the patch to avoid maintaining lowWatermark as state of Partition and compute it only on the fly. And `updateReplicaLogReadResult` would only compute LW if deleteRecordPurgatory is not empty.